value/binary.go

```diff
@@ -1149,17 +1149,63 @@ func init() {
 					if A.Rank() == 0 || B.Rank() == 0 {
 						Errorf("empty matrix for ,")
 					}
-					if A.Rank() != B.Rank()+1 || A.ElemSize() != B.Size() {
-						Errorf("catenate rank mismatch: %s != %s", NewIntVector(A.shape[1:]), NewIntVector(B.shape))
+					var shape []int
+					var data Vector
+					switch {
+					default:
+						Errorf("catenate shape mismatch: %s != %s", NewIntVector(A.shape[1:]), NewIntVector(B.shape))
```

Comment:
this is big enough i suggest we put it in a separate function

Reply:
Done.